src/types.js

```diff
@@ -0,0 +1,676 @@
+import type { Token } from "./tokenizer";
+import type { SourceLocation } from "./util/location";
+
+// Based on `ast/spec.md`.
+// But see "Not in spec" comments.
+
+export type Comment = {
+  type: "CommentBlock" | "CommentLine";
+  value: string;
+  start: number;
+  end: number;
+  loc: SourceLocation;
+};
+
+export interface NodeBase {
+  start: number;
+  end: number;
+  loc: SourceLocation;
+  range: [number, number];
+  leadingComments?: ?Array<Comment>;
```

Comment:
Flow has a [bug](https://flowtype.org/try/#0PQKgBAAgZgNg9gdzCYAoVBLAdgFwKYBOUAhgMZ5gAaYA3qmGDHsQCbYDmAwnALY964AzgC4wAfgCCBAsQCeAHm58BOAHwBuVAF90pOFkE4wAD1HUAvLXqNmbLF178hogNoBdbeqA) with maybe-typed properties in interfaces. It's better to avoid interfaces completely unless they only contain methods